mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp

```diff
@@ -1167,508 +970,138 @@ struct GridwiseGemmV2RewritePattern
-          loc,
-          b.create<DivUIOp>(loc, tid,
-                            GemmBBlockCopyClusterLengths_GemmNConstantOp),
-          GemmBBlockCopyClusterLengths_GemmKPackConstantOp);
-      GemmBBlockCopyThreadClusterId_Y = b.create<RemUIOp>(
-          loc, tid, GemmBBlockCopyClusterLengths_GemmNConstantOp);
-      GemmBBlockCopyThreadClusterId_X = b.create<RemUIOp>(
-          loc,
-          b.create<DivUIOp>(loc, tid,
-                            GemmBBlockCopyClusterLengths_GemmNConstantOp),
-          GemmBBlockCopyClusterLengths_GemmKPackConstantOp);
-
-      GemmBThreadDataIdBegin_Z =
-          b.create<MulIOp>(loc, GemmBBlockCopyThreadClusterId_Z,
-                           GemmBBlockCopyThreadSliceLengths_GemmKConstantOp);
-      GemmBThreadDataIdBegin_Y =
-          b.create<MulIOp>(loc, GemmBBlockCopyThreadClusterId_Y,
-                           GemmBBlockCopyThreadSliceLengths_GemmNConstantOp);
-      GemmBThreadDataIdBegin_X = b.create<MulIOp>(
-          loc, GemmBBlockCopyThreadClusterId_X,
-          GemmBBlockCopyThreadSliceLengths_GemmKPackConstantOp);
-    } else {
-      GemmBBlockCopyThreadClusterId_Y = b.create<DivUIOp>(
-          loc, tid, GemmBBlockCopyClusterLengths_GemmNConstantOp);
-      GemmBBlockCopyThreadClusterId_X = b.create<RemUIOp>(
-          loc, tid, GemmBBlockCopyClusterLengths_GemmNConstantOp);
-      GemmBThreadDataIdBegin_Y =
-          b.create<MulIOp>(loc, GemmBBlockCopyThreadClusterId_Y,
-                           GemmBBlockCopyThreadSliceLengths_GemmKConstantOp);
-      GemmBThreadDataIdBegin_X =
-          b.create<MulIOp>(loc, GemmBBlockCopyThreadClusterId_X,
-                           GemmBBlockCopyThreadSliceLengths_GemmNConstantOp);
-    }
+    TransformingForOp blockwiseLoadA =
+        createGlobalLoadLoop(b, loc, wrappedA, aVectorGlobalMap, aCopyPerThread,
+                             aVectorLen, bid, tid);
+    TransformingForOp blockwiseLoadB =
+        createGlobalLoadLoop(b, loc, wrappedB, bVectorGlobalMap, bCopyPerThread,
+                             bVectorLen, bid, tid);
 
-    Value GemmBBlockCopySourceCoord_Y;
-    Value GemmBBlockCopySourceCoord_X;
+    // Obtain XDLOPS-related attributes.
+    int64_t MPerWave = tuningParams.getMPerWave();
+    int64_t NPerWave = tuningParams.getNPerWave();
+    // int64_t MWaves = mPerBlock / MPerWave;
+    int64_t NWaves = nPerBlock / NPerWave;
 
-    Value GemmBBlockCopySourceCoord_Z;
-    if (KPack > 1) {
-      GemmBBlockCopySourceCoord_Z = GemmBThreadDataIdBegin_Z;
-      GemmBBlockCopySourceCoord_Y = b.create<AddIOp>(
-          loc, n_block_data_on_global, GemmBThreadDataIdBegin_Y);
-      GemmBBlockCopySourceCoord_X = GemmBThreadDataIdBegin_X;
-    } else {
-      GemmBBlockCopySourceCoord_Y = GemmBThreadDataIdBegin_Y;
-      GemmBBlockCopySourceCoord_X = b.create<AddIOp>(
-          loc, n_block_data_on_global, GemmBThreadDataIdBegin_X);
-    }
+    auto MPerWaveConstantOp = b.create<ConstantIndexOp>(loc, MPerWave);
```

Comment:
Yes, blockwise level ops isn't using the layout information as it should, it will be a future PR for sure.

On a side topic, it looks like this PR CI starts to catch all sorts of corner cases again just like last time (in the vectorization PR). I'll try to touch less things functionally and throw a nightly run at this PR before merging it.